internal/bminventory/inventory_v2_handlers.go

```diff
@@ -638,7 +689,29 @@ func (b *bareMetalInventory) infraEnvIPXEScript(ctx context.Context, infraEnv *c
 		return "", errors.Wrap(err, "failed to sign initrd URL")
 	}
 
-	return fmt.Sprintf(ipxeScriptFormat, initrdURL, kernelURL, rootfsURL), nil
+	return fmt.Sprintf(ipxeSecondScriptFormat, initrdURL, kernelURL, rootfsURL), nil
+}
+
+func (b *bareMetalInventory) secondIPXEScript(ctx context.Context, infraEnv *common.InfraEnv, mac *strfmt.MAC) (string, error) {
+	shouldServe, err := b.shouldServeScript(infraEnv, mac)
+	if err != nil {
+		return "", err
+	}
+	if !shouldServe {
+		return "", common.NewApiError(http.StatusNotFound, errors.New("IPXE booting skipped.  Going to boot from HD"))
+	}
+	return b.bootIPXEScript(ctx, infraEnv)
+}
+
+func (b *bareMetalInventory) infraEnvIPXEScript(ctx context.Context, infraEnv *common.InfraEnv, mac *strfmt.MAC, bootControl bool) (string, error) {
+	if mac == nil || len(*mac) == 0 {
+		if bootControl {
+			return b.firstIPXEScript(infraEnv)
+		} else {
+			return b.bootIPXEScript(ctx, infraEnv)
+		}
+	}
+	return b.secondIPXEScript(ctx, infraEnv, mac)
 }
```

Comment:
I think if you combined these the logic would read better:

```go
func (b *bareMetalInventory) infraEnvIPXEScript(ctx context.Context, infraEnv *common.InfraEnv, mac *strfmt.MAC, bootControl bool) (string, error) {
    if mac != nil && len(*mac) != 0 {
        if err := b.shouldServeHostIPXEScript(infraEnv, mac); err != nil {
            return "", err
        }
    }

    if bootControl {
        return b.hostRedirectIPXEScript(infraEnv)
    }
    return b.bootIPXEScript(ctx, infraEnv)
```

Also this example assumes some functions are renamed/refactored as mentioned in the other comments.

Reply:
OK